[cocoa] Restrict access to unpartitioned third-party cookies

sysrqb opened the pull request:
#### 2cc95f03a409847211d270cc46d4bddceb09d26f
<pre>
[cocoa] Restrict access to unpartitioned third-party cookies
<a href="https://bugs.webkit.org/show_bug.cgi?id=283540">https://bugs.webkit.org/show_bug.cgi?id=283540</a>
<a href="https://rdar.apple.com/140386768">rdar://140386768</a>

Reviewed by Sihui Liu.

In the current implementation of partitioned cookies, all cookies associated
with a domain are included by default (i.e., both partitioned and
unpartitioned). This change restricts access to unpartitioned cookies and only
allows access when the request should have &quot;storage access&quot;. An example is a
third party iframe that was granted storage access,

This patch adopts an experimental SPI on NSMutableURLRequest
_setAllowOnlyPartitionedCookies that controls whether or not unpartitioned
cookies should be included in a request.

When accessing cookies from JavaScript, by default we get all cookies
associated with a domain. We now take an additional step an filter all cookies
that are not in the specified partition, if the response should only included
partitioned cookies.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/cookies/resources/echo-http-and-dom-cookies-and-notify-done.py:

Add new query parameter

* LayoutTests/http/tests/cookies/resources/set-cookie-and-redirect-back.py:

Specify cookie path.

* LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html: Added.

New test

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:

Add build flag

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:

Declare new SPI.

* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::thirdPartyCookieBlockingDecisionForRequest const):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::cookiesForURL):
(WebCore::NetworkStorageSession::httpCookiesForURL const):
(WebCore::NetworkStorageSession::cookiesForURL const):
(WebCore::NetworkStorageSession::deleteCookie const):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::createWebSocketTask):

Restrict access to unpartitioned cookies, if necessary.

* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
(WebKit::CoreIPCNSURLRequest::CoreIPCNSURLRequest):
(WebKit::CoreIPCNSURLRequest::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:

Add new attribute.

Canonical link: <a href="https://commits.webkit.org/287558@main">https://commits.webkit.org/287558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b1727b1d0dfa8773ea49ccf5a0407ab46549ce01

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80078 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33485 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84603 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31064 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82187 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68140 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7374 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62613 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20440 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83146 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/52682 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72940 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42916 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50013 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27094 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29526 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73067 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71136 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27601 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86036 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79135 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7294 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5169 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/70895 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7471 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70128 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14116 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13066 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101543 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12388 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7258 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24754 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7100 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10618 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8907 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1905,6 +1905,15 @@ - (void)URLSession:(NSURLSession *)session webSocketTask:(NSURLSessionWebSocketT
             ensureMutableRequest()._privacyProxyFailClosedForUnreachableNonMainHosts = YES;
     }
 
+#if defined(CFN_HAS_ALLOW_ONLY_PARTITIONED_COOKIES)
```
Can we add a flag in PlatformHave.h? Maybe also specify the build numbers there (I assume we don't expect this to work for downlevel builds).

sysrqb replied:
I originally tried adding a flag in PlatformHave while still checking `CFN_HAS_ALLOW_ONLY_PARTITIONED_COOKIES`, but we don't include the CFNetwork header in that file (and I don't want to add it). Defining a `HAVE_ALLOW_ONLY_PARTITIONED_COOKIES` based on build numbers is a little sad, given we have this CFN macro, but I don't have a strong opinion about it and I can make this change. You're correct that this won't work on downlevels.

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1905,6 +1905,15 @@ - (void)URLSession:(NSURLSession *)session webSocketTask:(NSURLSessionWebSocketT
 
+#if defined(CFN_HAS_ALLOW_ONLY_PARTITIONED_COOKIES)
+    if ([mutableRequest respondsToSelector:@selector(_setAllowOnlyPartitionedCookies:)]) {
+        if (auto* storageSession = networkStorageSession(); storageSession && storageSession->isOptInCookiePartitioningEnabled()) {
```
`CheckedPtr storageSession`

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -270,16 +270,29 @@ - (CFURLRef)_cfurl;
-static RetainPtr<NSArray> cookiesForURL(NSHTTPCookieStorage *storage, NSURL *url, NSURL *mainDocumentURL, const std::optional<SameSiteInfo>& sameSiteInfo, NSString *partition = nullptr)
+static RetainPtr<NSArray> cookiesForURL(NSHTTPCookieStorage *storage, NSURL *url, NSURL *mainDocumentURL, const std::optional<SameSiteInfo>& sameSiteInfo, NSString *partition = nullptr, ThirdPartyCookieBlockingDecision thirdPartyCookieBlockingDecision = ThirdPartyCookieBlockingDecision::All)
 {
+    ASSERT(thirdPartyCookieBlockingDecision != ThirdPartyCookieBlockingDecision::All);
```
It does not make much sense that we don't think `ThirdPartyCookieBlockingDecision::All` is a valid value while the function uses that as default value.
Either this functions should not specify the default value, or the assertion should be removed.

sysrqb replied:
Good point.

szewai commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -162,19 +162,25 @@ ThirdPartyCookieBlockingDecision NetworkStorageSession::thirdPartyCookieBlocking
+    const auto decideThirdPartyCookieBlocking = [isOptInCookiePartitioningEnabled = isOptInCookiePartitioningEnabled()] (bool shouldAllowUnpartitionedCookies) {
+        if (shouldAllowUnpartitionedCookies)
+            return ThirdPartyCookieBlockingDecision::None;
+        return isOptInCookiePartitioningEnabled ? ThirdPartyCookieBlockingDecision::AllExceptPartitioned : ThirdPartyCookieBlockingDecision::All;
```
I think the enum value `AllExceptPartitioned` and `isOptInCookiePartitioningEnabled` flag might need to be protected by the compile time flag -- so it's only possible to set that value on available builds.

szewai commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -2804,6 +2804,9 @@ http/tests/storageAccess/has-storage-access-false-by-default.https.html [ Pass ]
 http/tests/storageAccess/has-storage-access-false-by-default-ephemeral.https.html [ Pass ]
 http/tests/storageAccess/has-storage-access-true-if-third-party-has-cookies.https.html [ Pass Timeout ]
 
+# Enable on appropriate platforms: rdar://140222322
```
Should we just skip it in `LayoutTests/TestExpectations`?

sysrqb replied:
No, I tried that, but we explicitly skip all `http/tests/storageAccess` tests in LayoutTests/TestExpectations, and then in `LayoutTests/platform/mac-wk2/TestExpectations` we set `http/tests/storageAccess [ Pass ]`, so we need to also explicitly skip this test. Skipping this test at a higher level is ignored. But, maybe we don't need this iOS exception.

szewai commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -162,19 +162,29 @@ ThirdPartyCookieBlockingDecision NetworkStorageSession::thirdPartyCookieBlocking
     if (hasStorageAccess(resourceDomain, firstPartyDomain, frameID, pageID))
         return ThirdPartyCookieBlockingDecision::None;
 
+    const auto decideThirdPartyCookieBlocking = [isOptInCookiePartitioningEnabled = isOptInCookiePartitioningEnabled()] (bool shouldAllowUnpartitionedCookies) {
```
It seems this should be under `#if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)` too, so is the function `isOptInCookiePartitioningEnabled()` (since the flag is only conditionalized). On the builds ALLOW_ONLY_PARTITIONED_COOKIES is false, we just return `ThirdPartyCookieBlockingDecision::All`.

sysrqb replied:
It's not necessary, but yes, we can do that.

szewai commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5443,6 +5443,7 @@ OptInPartitionedCookiesEnabled:
   category: dom
   humanReadableName: "Opt-in partitioned cookies (CHIPS)"
   humanReadableDescription: "Enable opt-in partitioned cookies"
+  condition: HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)
```
With this change, I think we need to move related functions (like`isOptInCookiePartitioningEnabled()`) under condition `if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)`

sysrqb replied:
We don't need to do that, it will default to false, but I made this change so it is clear.

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -270,16 +270,29 @@ - (CFURLRef)_cfurl;
+    if (!partition || thirdPartyCookieBlockingDecision == ThirdPartyCookieBlockingDecision::None)
+        return WTFMove(*cookiesPtr);
+
+    // Filter all unpartitioned cookies.
```
I don't quite understand this. Why will there be unpartitioned cookies in `cookiesPtr`?
If `partition` is not null, then `cookiesPtr` should only contain partitioned cookies.
If `partition` is null, it means the caller wants unpartitioned cookies. (Also you return early in the check above, so this code will not be invoked).

I am not sure why `cookiesForURL()` needs both `partition` and `thirdPartyCookieBlockingDecision`.

szewai replied:
Matthew explained offline that this is because `_getCookiesForURL` currently returns **only unpartitioned cookies** if partition is nil, and **unpartitioned cookies plus cookies in the specified partition** if partition is not nil.
So in the case of `ThirdPartyCookieBlockingDecision::AllExceptPartitioned`, we need to filter the unpartitioned cookies from result.

We may need a comment here.

Commit:
[cocoa] Restrict access to unpartitioned third-party cookies
https://bugs.webkit.org/show_bug.cgi?id=283540
rdar://140386768

Reviewed by Sihui Liu.

In the current implementation of partitioned cookies, all cookies associated
with a domain are included by default (i.e., both partitioned and
unpartitioned). This change restricts access to unpartitioned cookies and only
allows access when the request should have "storage access". An example is a
third party iframe that was granted storage access,

This patch adopts an experimental SPI on NSMutableURLRequest
_setAllowOnlyPartitionedCookies that controls whether or not unpartitioned
cookies should be included in a request.

When accessing cookies from JavaScript, by default we get all cookies
associated with a domain. We now take an additional step an filter all cookies
that are not in the specified partition, if the response should only included
partitioned cookies.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/cookies/resources/echo-http-and-dom-cookies-and-notify-done.py:

Add new query parameter

* LayoutTests/http/tests/cookies/resources/set-cookie-and-redirect-back.py:

Specify cookie path.

* LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/storageAccess/grant-with-prompt-under-general-third-party-cookie-blocking-with-partitioned-cookies.https.html: Added.

New test

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:

Add build flag

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:

Declare new SPI.

* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::thirdPartyCookieBlockingDecisionForRequest const):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::cookiesForURL):
(WebCore::NetworkStorageSession::httpCookiesForURL const):
(WebCore::NetworkStorageSession::cookiesForURL const):
(WebCore::NetworkStorageSession::deleteCookie const):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::createWebSocketTask):

Restrict access to unpartitioned cookies, if necessary.

* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
(WebKit::CoreIPCNSURLRequest::CoreIPCNSURLRequest):
(WebKit::CoreIPCNSURLRequest::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:

Add new attribute.

Canonical link: https://commits.webkit.org/287558@main
